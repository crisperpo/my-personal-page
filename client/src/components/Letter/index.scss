.letter {
    mix-blend-mode: color-burn;
    position: relative;
    transition: top ease 0.1s;
    transition-delay: 2s;
    transition: top ease 0.1s;
    transition-delay: 2s;
  
    &.hovered {
      animation: letterAnimation 3s;
    }
  }
  @keyframes letterAnimation {
    0% {
      top: 0;
    }
    50% {
      top: 0.2em;
    }
    100% {
      top: 0;
    }
  }
  