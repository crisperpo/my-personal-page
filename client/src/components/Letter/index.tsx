import { useEffect, useRef, useState } from 'react';

import { LetterContainer } from './index.styled';
import './index.scss';
import { useTheme } from 'styled-components';

const Letter = ({ letter, position }: { letter: string, position: number }) => {
    const [isHovered, setIsHovered] = useState(false);
    const letterRef = useRef<HTMLDivElement | null>(null);
    const theme = useTheme();
    useEffect(() => {
        const handleAnimationEnd = () => {
            setIsHovered(false);
        };
        const letter = letterRef.current;
        if (letter) {
            letter.addEventListener('animationend', handleAnimationEnd);
        }
        return () => {
            if (letter) {
                letter.removeEventListener('animationend', handleAnimationEnd);
            }
        };
    }, []);
    const colorKeys = Object.values(theme.title);
    const color = colorKeys[position % colorKeys.length];

    const handleOnHover = () => {
        setIsHovered(true);
    };

    const letterClass = isHovered ? 'letter hovered' : 'letter';

    return (
        <LetterContainer
            className={letterClass}
            color={color}
            onMouseEnter={handleOnHover}
            onClick={handleOnHover}
            ref={letterRef}
        >
            {letter}
        </LetterContainer>
    );
};

export default Letter;
