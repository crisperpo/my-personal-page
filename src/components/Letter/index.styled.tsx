import styled from 'styled-components';
import colors from '../../lib/constants/colors';

const StyledLetterContainer = styled.div<{ color: string }>`
    color: ${(props) => props.color};
    mix-blend-mode: color-burn;
`;

const LetterContainer = ({position, children}: {position: number; children: React.ReactNode}) => {
    const colorKeys = Object.values(colors.title);
    const color = colorKeys[position % colorKeys.length];
    return (
        <StyledLetterContainer color={color}>{children}</StyledLetterContainer>
    );
};

export {
    LetterContainer
};
