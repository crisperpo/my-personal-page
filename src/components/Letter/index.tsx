import { useEffect, useRef, useState } from 'react';
import { LetterContainer } from './index.styled';
import colors from '@/lib/constants/colors';
import './index.css'

const Letter = ({ letter, position }: { letter: string, position: number }) => {
    const [isHovered, setIsHovered] = useState(false);
    const letterRef = useRef<HTMLDivElement | null>(null);
    useEffect(() => {
        const handleAnimationEnd = () => {
            setIsHovered(false);
        };
        const letter = letterRef.current;
        if (letter) {
            letter.addEventListener('animationend', handleAnimationEnd);
        }
        return () => {
            if (letter) {
                letter.removeEventListener('animationend', handleAnimationEnd);
            }
        };
    }, []);
    const colorKeys = Object.values(colors.title);
    const color = colorKeys[position % colorKeys.length];

    const handleOnHover = () => {
        setIsHovered(true);
    };

    const letterClass = isHovered ? 'letter hovered' : 'letter';

    return (
        <LetterContainer
            className={letterClass}
            color={color}
            onMouseEnter={handleOnHover}
            onClick={handleOnHover}
            ref={letterRef}
        >
            {letter}
        </LetterContainer>
    );
};

export default Letter;
