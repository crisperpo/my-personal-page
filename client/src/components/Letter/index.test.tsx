import Letter from './index';
import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';

import themes from '../../lib/constants/theme';

// Mocks
vi.mock('../../hooks/useTheme', () => ({
    default: vi.fn().mockReturnValue(themes.light)
}));

const MOCK_LETTER = 'A';

describe('Letter Component', () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });
    it('should render correctly', () => {
        render(
            <Letter letter={MOCK_LETTER} position={0} />
        );
    });
    it('should contain the class "letter"', () => {
        render(
            <Letter letter={MOCK_LETTER} position={0} />
        );
        const letterElement = screen.getByText(MOCK_LETTER).closest('div');
        expect(letterElement).toHaveClass('letter');
    });
    it('should contain the class "hovered" when isHovered is true', () => {
        vi.mock('react', async () => {
            const actualReact = await vi.importActual('react');
            return {
                ...actualReact,
                useState: vi.fn().mockImplementation(() => [true, vi.fn()]),
            };
        });

        render(<Letter letter={MOCK_LETTER} position={0} />);
        const letterElement = screen.getByText(MOCK_LETTER).closest('div');
        expect(letterElement).toHaveClass('hovered');
    });
    it('should render a div containing the letter with the correct color', () => {
        const MOCK_POSITION = 1;
        const colorKeys = Object.values(themes.light.title);
        const color = colorKeys[MOCK_POSITION % colorKeys.length];
        render(
            <Letter letter={MOCK_LETTER} position={MOCK_POSITION} />
        );
        const letterDiv = screen.getByText(MOCK_LETTER).closest('div');
        expect(letterDiv).toHaveStyle(`color: ${color}`);
    });
    it('matches the snapshot', () => {
        const { asFragment } = render(
            <Letter letter={MOCK_LETTER} position={0} />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
