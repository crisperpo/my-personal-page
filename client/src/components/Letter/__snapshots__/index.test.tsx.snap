// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Letter Component > matches the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  color: #E21E25;
  mix-blend-mode: color-burn;
  transition: top ease 0.1s;
  transition-delay: 2s;
  position: relative;
}

.c0:hovered {
  animation: emeSZq 3s;
}

<div
    class="c0 letter hovered"
    color="#E21E25"
  >
    A
  </div>
</DocumentFragment>
`;
